src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs

```diff
@@ -0,0 +1,444 @@
+    {
+        private Stream baseStream;
+        private long baseStreamPosition = 0;  // virtual Position of the base stream.
+        private long maxAllowedPosition = long.MaxValue;
+        private int bufferCapacity;
+        private byte[] buffer;
+        private int bufferOffset = 0;
+        private int bufferLength = 0;
+        private Encoding encoding;
+
+        public BufferedBinaryReader(Stream stream, Encoding? encoding = null, int bufferCapacity = 32768)
+        {
+            if (!stream.CanRead)
+            {
+                throw new InvalidOperationException(ResourceUtilities.GetResourceString("Binlog_StreamUtils_MustBeReadable"));
+            }
+
+            baseStream = stream;
+            this.bufferCapacity = bufferCapacity;  // Note: bufferSize must be large enough for an Read operation.
+            this.encoding = encoding ?? new UTF8Encoding();
+            buffer = new byte[this.bufferCapacity];
+        }
+
+        /// <summary>
+        /// Position of the base stream.
+        /// </summary>
+        public long Position => baseStreamPosition;
+
+        /// <summary>
+        /// Number of bytes allowed to read.  If set, then read functions will throw if exceeded by the amount.
+        /// </summary>
+        public int? BytesCountAllowedToRead
+        {
+            set
+            {
+                if (value.HasValue)
+                {
+                    if (value.Value < 0)
+                    {
+                        throw new Exception();
+                    }
+
+                    maxAllowedPosition = baseStreamPosition + value.Value;
+                }
+                else
+                {
+                    maxAllowedPosition = long.MaxValue;
+                }
+            }
+        }
+
+        /// <summary>
+        /// If <see cref="BytesCountAllowedToRead"/> is set, then this is the number of bytes remaining to read.  Otherwise, 0.
+        /// </summary>
+        public int BytesCountAllowedToReadRemaining => maxAllowedPosition == long.MaxValue ? 0 : (int)(maxAllowedPosition - baseStreamPosition);
+
+        /// <summary>
+        /// Reads a 32-bit signed integer.
+        /// </summary>
+        /// <returns>Return a integer.</returns>
+        public int ReadInt32()
+        {
+            FillBuffer(4);
+
+            var result = (int)(buffer[bufferOffset] | buffer[bufferOffset + 1] << 8 | buffer[bufferOffset + 2] << 16 | buffer[bufferOffset + 3] << 24);
+            bufferOffset += 4;
+            baseStreamPosition += 4;
+            return result;
+        }
+
+        // Reusable StringBuilder for ReadString().
+        private StringBuilder? cachedBuilder;
+
+        // Reusable char[] for ReadString().
+        private char[]? charBuffer;
+
+        /// <summary>
+        /// Reads a string with a prefixed of the length.
+        /// </summary>
+        /// <returns>A string.</returns>
+        public string ReadString()
```

Comment:
Where is this implementation comming from? Is it adapted from the BCL BinaryReader? We should have that info in comments. I already saw couple cases where MSBuild code had some custom implementation copied from BCL - while the BCL diverged a bit over time and nobody knew the original motivation or sometime even the source

Reply:
ReadString() is effectively re-written when compared to BCL.  Should review in full.

Reply:
I modified to look closer to the version in BCL.  It is very similar, I moved the one time branch outside of the loop to simplified the loop.